<div style="text-align:center;margin:40px 0;">
    <h1 style="font-weight:bolder;color:#283d3b;">Rock and Mine Prediction</h1>
</div>

<div style="margin:40px 0;">
    <h2 style="font-weight:bolder;color:#197278;">Importing Libraries</h2>
</div>

In [1]:
import numpy as np
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score

C:\Users\PMLS\AppData\Local\Temp\ipykernel_7136\3120435754.py:2: DeprecationWarning: 
Pyarrow will become a required dependency of pandas in the next major release of pandas (pandas 3.0),
(to allow more performant data types, such as the Arrow string type, and better interoperability with other libraries)
but was not found to be installed on your system.
If this would cause problems for you,
please provide us feedback at https://github.com/pandas-dev/pandas/issues/54466
        
  import pandas as pd


<div style="margin:40px 0;">
    <h2 style="font-weight:bolder;color:#197278;">Importing Data</h2>
</div>

In [2]:
data = pd.read_csv('sonar data.csv', header=None)

<div style="margin:40px 0;">
    <h2 style="font-weight:bolder;color:#197278;">Data</h2>
</div>

In [3]:
data.head()

0       1       2       3       4       5       6       7       8   \
0  0.0200  0.0371  0.0428  0.0207  0.0954  0.0986  0.1539  0.1601  0.3109   
1  0.0453  0.0523  0.0843  0.0689  0.1183  0.2583  0.2156  0.3481  0.3337   
2  0.0262  0.0582  0.1099  0.1083  0.0974  0.2280  0.2431  0.3771  0.5598   
3  0.0100  0.0171  0.0623  0.0205  0.0205  0.0368  0.1098  0.1276  0.0598   
4  0.0762  0.0666  0.0481  0.0394  0.0590  0.0649  0.1209  0.2467  0.3564   

       9   ...      51      52      53      54      55      56      57  \
0  0.2111  ...  0.0027  0.0065  0.0159  0.0072  0.0167  0.0180  0.0084   
1  0.2872  ...  0.0084  0.0089  0.0048  0.0094  0.0191  0.0140  0.0049   
2  0.6194  ...  0.0232  0.0166  0.0095  0.0180  0.0244  0.0316  0.0164   
3  0.1264  ...  0.0121  0.0036  0.0150  0.0085  0.0073  0.0050  0.0044   
4  0.4459  ...  0.0031  0.0054  0.0105  0.0110  0.0015  0.0072  0.0048   

       58      59  60  
0  0.0090  0.0032   R  
1  0.0052  0.0044   R  
2  0.0095  0.0078   R  
3  0.0040  0.0117   R  
4  0.0107  0.0094   R  

[5 rows x 61 columns]

<div style="margin:40px 0;">
    <h2 style="font-weight:bolder;color:#197278;">Shape of Data</h2>
</div>

In [4]:
data.shape

(208, 61)

<div style="margin:20px 0;background-color:#edddd4;padding:20px">
    Number of Columns: 61</br>
    Number of Rows: 208
</div>

<div style="margin:40px 0;">
    <h2 style="font-weight:bolder;color:#197278;">Data Description</h2>
</div>

In [5]:
data.describe()

0           1           2           3           4           5   \
count  208.000000  208.000000  208.000000  208.000000  208.000000  208.000000   
mean     0.029164    0.038437    0.043832    0.053892    0.075202    0.104570   
std      0.022991    0.032960    0.038428    0.046528    0.055552    0.059105   
min      0.001500    0.000600    0.001500    0.005800    0.006700    0.010200   
25%      0.013350    0.016450    0.018950    0.024375    0.038050    0.067025   
50%      0.022800    0.030800    0.034300    0.044050    0.062500    0.092150   
75%      0.035550    0.047950    0.057950    0.064500    0.100275    0.134125   
max      0.137100    0.233900    0.305900    0.426400    0.401000    0.382300   

               6           7           8           9   ...          50  \
count  208.000000  208.000000  208.000000  208.000000  ...  208.000000   
mean     0.121747    0.134799    0.178003    0.208259  ...    0.016069   
std      0.061788    0.085152    0.118387    0.134416  ...    0.012008   
min      0.003300    0.005500    0.007500    0.011300  ...    0.000000   
25%      0.080900    0.080425    0.097025    0.111275  ...    0.008425   
50%      0.106950    0.112100    0.152250    0.182400  ...    0.013900   
75%      0.154000    0.169600    0.233425    0.268700  ...    0.020825   
max      0.372900    0.459000    0.682800    0.710600  ...    0.100400   

               51          52          53          54          55          56  \
count  208.000000  208.000000  208.000000  208.000000  208.000000  208.000000   
mean     0.013420    0.010709    0.010941    0.009290    0.008222    0.007820   
std      0.009634    0.007060    0.007301    0.007088    0.005736    0.005785   
min      0.000800    0.000500    0.001000    0.000600    0.000400    0.000300   
25%      0.007275    0.005075    0.005375    0.004150    0.004400    0.003700   
50%      0.011400    0.009550    0.009300    0.007500    0.006850    0.005950   
75%      0.016725    0.014900    0.014500    0.012100    0.010575    0.010425   
max      0.070900    0.039000    0.035200    0.044700    0.039400    0.035500   

               57          58          59  
count  208.000000  208.000000  208.000000  
mean     0.007949    0.007941    0.006507  
std      0.006470    0.006181    0.005031  
min      0.000300    0.000100    0.000600  
25%      0.003600    0.003675    0.003100  
50%      0.005800    0.006400    0.005300  
75%      0.010350    0.010325    0.008525  
max      0.044000    0.036400    0.043900  

[8 rows x 60 columns]

<div style="margin:40px 0;">
    <h2 style="font-weight:bolder;color:#197278;">Checking for Null Values</h2>
</div>

In [6]:
data.isnull().sum().sum()

0

<div style="margin:40px 0;">
    <h2 style="font-weight:bolder;color:#197278;">Checking for Duplicate Values</h2>
</div>

In [7]:
data.duplicated().sum()

0

<div style="margin:40px 0;">
    <h2 style="font-weight:bolder;color:#197278;">Checking for ratio between Rocks and Mines</h2>
</div>

In [8]:
data.value_counts(60)

60
M    111
R     97
Name: count, dtype: int64

In [9]:
data[60].value_counts()

60
M    111
R     97
Name: count, dtype: int64

<div style="margin:20px 0;background-color:#edddd4;padding:20px">
    There is not a significant difference between the distribution of Mines and Rocks.
</div>

<div style="margin:40px 0;">
    <h2 style="font-weight:bolder;color:#197278;">Separation of features and columns</h2>
</div>

In [10]:
x = data.drop(columns=60,axis=1)
y = data[60]

In [11]:
x.shape

(208, 60)

In [12]:
y.shape

(208,)

In [13]:
y

0      R
1      R
2      R
3      R
4      R
      ..
203    M
204    M
205    M
206    M
207    M
Name: 60, Length: 208, dtype: object

<div style="margin:40px 0;">
    <h2 style="font-weight:bolder;color:#197278;">Separation of testing and training data</h2>
</div>

In [14]:
xtrain,xtest,ytrain,ytest = train_test_split(x,y,test_size=0.1,stratify=y,random_state=1)

In [15]:
xtrain.shape

(187, 60)

In [16]:
xtest.shape

(21, 60)

In [17]:
ytrain.shape

(187,)

In [18]:
ytest.shape

(21,)

<div style="margin:40px 0;">
    <h2 style="font-weight:bolder;color:#197278;">Model Development</h2>
</div>

In [19]:
model = LogisticRegression()

In [20]:
model.fit(xtrain,ytrain)

LogisticRegression()

<div style="margin:40px 0;">
    <h2 style="font-weight:bolder;color:#197278;">Accuracy Score of training data</h2>
</div>

In [21]:
print(accuracy_score(model.predict(xtrain), ytrain))

0.8342245989304813


<div style="margin:40px 0;">
    <h2 style="font-weight:bolder;color:#197278;">Accuracy Score of test data</h2>
</div>

In [22]:
print(accuracy_score(model.predict(xtest), ytest))

0.7619047619047619


<div style="margin:40px 0;">
    <h2 style="font-weight:bolder;color:#197278;">Model Prediction Tests</h2>
</div>

In [23]:
input = np.asarray((0.0192,0.0607,0.0378,0.0774,0.1388,0.0809,0.0568,0.0219,0.1037,0.1186,0.1237,0.1601,0.3520,0.4479,0.3769,0.5761,0.6426,0.6790,0.7157,0.5466,0.5399,0.6362,0.7849,0.7756,0.5780,0.4862,0.4181,0.2457,0.0716,0.0613,0.1816,0.4493,0.5976,0.3785,0.2495,0.5771,0.8852,0.8409,0.3570,0.3133,0.6096,0.6378,0.2709,0.1419,0.1260,0.1288,0.0790,0.0829,0.0520,0.0216,0.0360,0.0331,0.0131,0.0120,0.0108,0.0024,0.0045,0.0037,0.0112,0.0075))
input = input.reshape(1, -1)
print(model.predict(input))

['R']


In [24]:
input = np.asarray((0.0530,0.0885,0.1997,0.2604,0.3225,0.2247,0.0617,0.2287,0.0950,0.0740,0.1610,0.2226,0.2703,0.3365,0.4266,0.4144,0.5655,0.6921,0.8547,0.9234,0.9171,1.0000,0.9532,0.9101,0.8337,0.7053,0.6534,0.4483,0.2460,0.2020,0.1446,0.0994,0.1510,0.2392,0.4434,0.5023,0.4441,0.4571,0.3927,0.2900,0.3408,0.4990,0.3632,0.1387,0.1800,0.1299,0.0523,0.0817,0.0469,0.0114,0.0299,0.0244,0.0199,0.0257,0.0082,0.0151,0.0171,0.0146,0.0134,0.0056))
input = input.reshape(1, -1)
print(model.predict(input))

['M']


In [25]:
input = np.asarray((0.0162,0.0041,0.0239,0.0441,0.0630,0.0921,0.1368,0.1078,0.1552,0.1779,0.2164,0.2568,0.3089,0.3829,0.4393,0.5335,0.5996,0.6728,0.7309,0.8092,0.8941,0.9668,1.0000,0.9893,0.9376,0.8991,0.9184,0.9128,0.7811,0.6018,0.3765,0.3300,0.2280,0.0212,0.1117,0.1788,0.2373,0.2843,0.2241,0.2715,0.3363,0.2546,0.1867,0.2160,0.1278,0.0768,0.1070,0.0946,0.0636,0.0227,0.0128,0.0173,0.0135,0.0114,0.0062,0.0157,0.0088,0.0036,0.0053,0.0030))
input = input.reshape(1, -1)
print(model.predict(input))

['M']


In [26]:
input = np.asarray((0.0408,0.0653,0.0397,0.0604,0.0496,0.1817,0.1178,0.1024,0.0583,0.2176,0.2459,0.3332,0.3087,0.2613,0.3232,0.3731,0.4203,0.5364,0.7062,0.8196,0.8835,0.8299,0.7609,0.7605,0.8367,0.8905,0.7652,0.5897,0.3037,0.0823,0.2787,0.7241,0.8032,0.8050,0.7676,0.7468,0.6253,0.1730,0.2916,0.5003,0.5220,0.4824,0.4004,0.3877,0.1651,0.0442,0.0663,0.0418,0.0475,0.0235,0.0066,0.0062,0.0129,0.0184,0.0069,0.0198,0.0199,0.0102,0.0070,0.0055))
input = input.reshape(1, -1)
print(model.predict(input))

['R']


In [27]:
input = np.asarray((0.0240,0.0218,0.0324,0.0569,0.0330,0.0513,0.0897,0.0713,0.0569,0.0389,0.1934,0.2434,0.2906,0.2606,0.3811,0.4997,0.3015,0.3655,0.6791,0.7307,0.5053,0.4441,0.6987,0.8133,0.7781,0.8943,0.8929,0.8913,0.8610,0.8063,0.5540,0.2446,0.3459,0.1615,0.2467,0.5564,0.4681,0.0979,0.1582,0.0751,0.3321,0.3745,0.2666,0.1078,0.1418,0.1687,0.0738,0.0634,0.0144,0.0226,0.0061,0.0162,0.0146,0.0093,0.0112,0.0094,0.0054,0.0019,0.0066,0.0023))
input = input.reshape(1, -1)
print(model.predict(input))

['R']
